<a href="https://colab.research.google.com/github/Shashankkota/parkinson-s-disease-prediction/blob/main/Parkinsons_disease.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1)Make necessary imports:

Detecting Parkinson’s Disease with **XGBoost**

In [ ]:
import numpy as np
import pandas as pd
import os, sys
from sklearn.preprocessing import MinMaxScaler
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

2)  Now lets import the parkinsons disease dataset from local storage(Source:UCI ML Parkinsons dataset) and then Read the data into a DataFrame and get the first 5 records

In [ ]:
from google.colab import files
import pandas as pd

# Upload the file
uploaded = files.upload()

# Get the filename (assuming only one file is uploaded)
filename = list(uploaded.keys())[0]

# Read the file into a DataFrame
df = pd.read_csv(filename)

# Display the first few rows
df.head()


Saving parkinsons.data to parkinsons.data


name  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0  phon_R01_S01_1      119.992       157.302        74.997         0.00784   
1  phon_R01_S01_2      122.400       148.650       113.819         0.00968   
2  phon_R01_S01_3      116.682       131.111       111.555         0.01050   
3  phon_R01_S01_4      116.676       137.871       111.366         0.00997   
4  phon_R01_S01_5      116.014       141.781       110.655         0.01284   

   MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  ...  \
0           0.00007   0.00370   0.00554     0.01109       0.04374  ...   
1           0.00008   0.00465   0.00696     0.01394       0.06134  ...   
2           0.00009   0.00544   0.00781     0.01633       0.05233  ...   
3           0.00009   0.00502   0.00698     0.01505       0.05492  ...   
4           0.00011   0.00655   0.00908     0.01966       0.06425  ...   

   Shimmer:DDA      NHR     HNR  status      RPDE       DFA   spread1  \
0      0.06545  0.02211  21.033       1  0.414783  0.815285 -4.813031   
1      0.09403  0.01929  19.085       1  0.458359  0.819521 -4.075192   
2      0.08270  0.01309  20.651       1  0.429895  0.825288 -4.443179   
3      0.08771  0.01353  20.644       1  0.434969  0.819235 -4.117501   
4      0.10470  0.01767  19.649       1  0.417356  0.823484 -3.747787   

    spread2        D2       PPE  
0  0.266482  2.301442  0.284654  
1  0.335590  2.486855  0.368674  
2  0.311173  2.342259  0.332634  
3  0.334147  2.405554  0.368975  
4  0.234513  2.332180  0.410335  

[5 rows x 24 columns]

3)Extract the features and labels from the dataset. The features include all columns except ‘status’, while the labels correspond to the values in the ‘status’ column.

In [ ]:
#DataFlair - Get the features and labels
features=df.loc[:,df.columns!='status'].values[:,1:]
labels=df.loc[:,'status'].values

4)The ‘status’ column contains labels 0 and 1. Let's determine the count of each label to understand their distribution.

In [ ]:
print(labels[labels==1].shape[0], labels[labels==0].shape[0])

147 48


Our dataset contains 147 instances labeled as 1 and 48 labeled as 0 in the ‘status’ column.  

5)initialize a `MinMaxScaler` to normalize the features within a range of -1 to 1. This scaler adjusts the feature values proportionally within the specified range. The `fit_transform()` method is used to first compute the necessary scaling parameters and then apply the transformation. The labels remain unchanged as they do not require scaling.

In [ ]:
scaler=MinMaxScaler((-1,1))
x=scaler.fit_transform(features)
y=labels

6)Next, divide the dataset into training and testing sets, allocating 20% of the data for testing. This ensures the model is trained on the majority of the data while retaining a portion for evaluation.

In [ ]:
x_train,x_test,y_train,y_test=train_test_split(x, y, test_size=0.2, random_state=7)

7)Create an instance of XGBClassifier and train the model. This classifier utilizes eXtreme Gradient Boosting, a powerful gradient boosting algorithm designed for modern data science challenges. It belongs to the Ensemble Learning category in machine learning, where multiple models are trained and combined to generate a more accurate and robust prediction.

In [ ]:
model=XGBClassifier()
model.fit(x_train,y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=3, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=100, n_jobs=None,
              num_parallel_tree=None, random_state=42, ...)

Generate `y_pred`, which contains the predicted values for `x_test`, and evaluate the model’s accuracy. Finally, print the accuracy score to assess the model's performance.

In [ ]:
y_pred=model.predict(x_test)
print(accuracy_score(y_test, y_pred)*100)

94.87179487179486


In the given Parkinson’s Disease Detection Project, the model is trained on a dataset, but there is no direct user input to predict whether a patient has the disease.

Now we need to add User Input for Prediction
To allow user input (e.g., from a form or console), you need to:

Collect the required features from the user.
Preprocess the input like the training data.
Use the trained XGBClassifier model to predict if the patient has Parkinson’s disease.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

df = pd.read_csv('parkinsons.data')

X = df.drop(columns=['name', 'status'])
y = df['status']

scaler = MinMaxScaler((-1, 1))
X_scaled = scaler.fit_transform(X)

x_train, x_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

model = XGBClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)
model.fit(x_train, y_train)

def predict_parkinsons():
    print("Enter the following 22 values:")

    user_input = []
    feature_names = df.columns.drop(['name', 'status'])

    for feature in feature_names:
        value = float(input(f"{feature}: "))
        user_input.append(value)

    user_input = np.array(user_input).reshape(1, -1)
    user_input = scaler.transform(user_input)
    prediction = model.predict(user_input)

    if prediction[0] == 1:
        print("The model predicts that the patient has Parkinson’s Disease.")
    else:
        print("The model predicts that the patient does NOT have Parkinson’s Disease.")

predict_parkinsons()


FileNotFoundError: [Errno 2] No such file or directory: 'parkinsons.data'